src/nomad_simulations/schema_packages/model_system.py

```diff
@@ -1150,46 +1303,94 @@ def resolve_system_type_and_dimensionality(
 
         return system_type, dimensionality
 
+    def populate_atom_types(self, logger: 'BoundLogger') -> None:
+        """
+        Populates the `atom_types` subsection with unique AtomDefn entries
+        extracted from the `atom_states` subsection.
+        """
+        unique_types = {}
+        for state in self.atom_states:
+            atom_def = state.atom_definition_ref
```

Comment:
i am not sure i follow, you want to populate atom_types but you read them from state.atom_definition_ref but this also references the one in atom_types?